### ISSUE TITLE
With py.test raises on python2.6, weird exception instances

### ISSUE THREAD

**Issue by alex** [pytest-dev/pytest#988]
I don't know how to describe this, so here's a minimal test case:

``` py
import errno
import socket

import pytest


def test_foo():
    s = socket.socket()
    with pytest.raises(socket.error) as cm:
        s.connect(("127.0.0.1", 1))
    assert cm.value.errno == errno.ECONNREFUSED
```

and this fails with:

```
====================================================== test session starts =======================================================
platform darwin -- Python 2.6.9 -- py-1.4.30 -- pytest-2.7.2
rootdir: /Users/alex_gaynor/.virtualenvs/tempenv-39fb22646f9f, inifile:
collected 1 items

test.py F

============================================================ FAILURES ============================================================
____________________________________________________________ test_foo ____________________________________________________________

    def test_foo():
        s = socket.socket()
        with pytest.raises(socket.error) as cm:
            s.connect(("127.0.0.1", 1))
>       assert cm.value.errno == errno.ECONNREFUSED
E       assert None == 61
E        +  where None = error((61, 'Connection refused'),).errno
E        +    where error((61, 'Connection refused'),) = <ExceptionInfo error tblen=2>.value
E        +  and   61 = errno.ECONNREFUSED

test.py:11: AssertionError
==================================================== 1 failed in 0.01 seconds ====================================================
```

if you just use a normal `try` / `except` you get an exception with a correct `errno`

**Comment by alex** [comment]
`--assert=plain` does not fix the bug

**Comment by alex** [comment]
here is an expanded test which shows that the try/except works fine.

```py
import errno
import socket

import pytest


def test_foo():
    s = socket.socket()
    with pytest.raises(socket.error) as cm:
        s.connect(("127.0.0.1", 1))
    assert cm.value.errno == errno.ECONNREFUSED

def test_bar():
    s = socket.socket()
    try:
        s.connect(("127.0.0.1", 1))
    except socket.error as e:
        pass
    assert e.errno == errno.ECONNREFUSED
```

**Comment by RonnyPfannschmidt** [comment]
how is the behavior with `cm = pytest.raises(socket.error, s.connect, ("127.0.0.1", 1))`

**Comment by alex** [comment]
@RonnyPfannschmidt passes:

```
(tempenv-2040140165f5f) ~/.v/tempenv-2040140165f5f $ cat test.py
import errno
import socket

import pytest


def test_foo():
    s = socket.socket()
    cm = pytest.raises(socket.error, s.connect, ("127.0.0.1", 1))
    assert cm.value.errno == errno.ECONNREFUSED

(tempenv-2040140165f5f) ~/.v/tempenv-2040140165f5f $ py.test test.py
====================================================== test session starts =======================================================
platform darwin -- Python 2.6.9 -- py-1.4.30 -- pytest-2.7.2
rootdir: /Users/alex_gaynor/.virtualenvs/tempenv-2040140165f5f, inifile:
collected 1 items

test.py .

==================================================== 1 passed in 0.01 seconds ====================================================
```

**Comment by nicoddemus** [comment]
Hmm this might be related or similar to #767.

I can reproduce this problem on master.

**Comment by RonnyPfannschmidt** [comment]
i suspect the ExceptionInfo initialization hack is bogous and it shows on python2.6

its reasonably easy to fix after moving py.code into py.test (however moving py.code into py.test is a major undertaking)

**Comment by RonnyPfannschmidt** [comment]
as a workaround we could perhaps use a new DeferedExceptionInfo class which shares the api with exceptioninfo, but is initialized without a exception, and adds the exception details later

**Comment by The-Compiler** [comment]
While this should be fixed, it shouldn't block 3.0. Changing the milestone to 3.1 for now.

**Comment by RonnyPfannschmidt** [comment]
#2822  would close this one
